File: implementation/scipamato/core/web/src/test/java/ch/difty/scipamato/core/web/paper/common/SearchablePaperPanelTest.java
package ch.difty.scipamato.core.web.paper.common;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.isA;
import static org.mockito.Mockito.*;

import de.agilecoders.wicket.core.markup.html.bootstrap.button.BootstrapButton;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.model.Model;
import org.junit.jupiter.api.Test;

import ch.difty.scipamato.common.persistence.paging.PaginationContext;
import ch.difty.scipamato.core.entity.search.PaperFilter;
import ch.difty.scipamato.core.entity.search.SearchCondition;

@SuppressWarnings("SpellCheckingInspection")
class SearchablePaperPanelTest extends PaperPanelTest<SearchCondition, SearchablePaperPanel> {

    @Override
    protected SearchablePaperPanel makePanel() {
        SearchCondition sc = new SearchCondition();
        sc.setId("1");
        sc.setNumber("100");
        sc.setAuthors("a");
        sc.setFirstAuthor("fa");
        sc.setFirstAuthorOverridden(false);

        sc.setTitle("t");
        sc.setLocation("l");
        sc.setPublicationYear("2017");

        sc.setPmId("pmid");
        sc.setDoi("doi");
        sc.setCreatedDisplayValue("cdv");
        sc.setModifiedDisplayValue("lmdv");

        sc.setGoals("g");
        sc.setPopulation("p");
        sc.setMethods("m");

        sc.setPopulationPlace("ppl");
        sc.setPopulationParticipants("ppa");
        sc.setPopulationDuration("pd");
        sc.setExposurePollutant("ep");
        sc.setExposureAssessment("ea");
        sc.setMethodStudyDesign("msd");
        sc.setMethodOutcome("mo");
        sc.setMethodStatistics("ms");
        sc.setMethodConfounders("mc");

        sc.setResult("r");
        sc.setIntern("i");
        sc.setResultMeasuredOutcome("rmo");
        sc.setResultExposureRange("rer");
        sc.setResultEffectEstimate("ree");
        sc.setConclusion("cc");
        sc.setComment("c");

        sc.addCode(newC(1, "F"));
        sc.setMainCodeOfCodeclass1("mcocc1");
        sc.addCode(newC(2, "A"));
        sc.addCode(newC(3, "A"));
        sc.addCode(newC(4, "A"));
        sc.addCode(newC(5, "A"));
        sc.addCode(newC(6, "A"));
        sc.addCode(newC(7, "A"));
        sc.addCode(newC(8, "A"));

        sc.setOriginalAbstract("oa");

        return new SearchablePaperPanel("panel", Model.of(sc)) {
            private static final long serialVersionUID = 1L;

            @Override
            protected void onFormSubmit() {
                // no-op
            }

            @Override
            protected void restartSearchInPaperSearchPage() {
                // no-op
            }

            @Override
            protected void doOnSubmit() {
                // no-op
            }
        };
    }

    @Override
    protected void tearDownLocalHook() {
        verify(paperServiceMock, times(2)).findPageOfIdsByFilter(isA(PaperFilter.class), isA(PaginationContext.class));
        verifyNoMoreInteractions(paperServiceMock);
    }

    @Override
    protected void assertSpecificComponents() {
        String b = "panel";
        getTester().assertComponent(b, SearchablePaperPanel.class);

        assertCommonComponents(b);

        b += ":form";
        assertTextFieldWithLabel(b + ":id", "1", "ID");
        assertTextFieldWithLabel(b + ":number", "100", "SciPaMaTo-Core-No.");
        assertTextFieldWithLabel(b + ":publicationYear", "2017", "Pub. Year");
        assertTextFieldWithLabel(b + ":pmId", "pmid", "PMID");
        getTester().assertLabel(b + ":submit:label", "Search");
        assertTextFieldWithLabel(b + ":createdDisplayValue", "cdv", "Created");
        assertTextFieldWithLabel(b + ":modifiedDisplayValue", "lmdv", "Last Modified");

        getTester().assertComponent(b + ":submit", BootstrapButton.class);
        verifyCodeAndCodeClassCalls(1);
    }

    @Test
    void specificFields_areEnabled() {
        getTester().startComponentInPage(makePanel());
        getTester().isEnabled("panel:form:id");
        getTester().isEnabled("panel:form:number");
        getTester().isEnabled("panel:form:firstAuthorOverridden");
        getTester().isEnabled("panel:form:createdDisplayValue");
        getTester().isEnabled("panel:form:modifiedDisplayValue");
    }

    @Test
    void summary_doesNotExist() {
        getTester().startComponentInPage(makePanel());
        getTester().assertContainsNot("panel:form:summary");
    }

    @Test
    void summaryShort_doesNotExist() {
        getTester().startComponentInPage(makePanel());
        getTester().assertContainsNot("panel:form:summaryShort");
    }

    @Test
    void navigationButtons_andPubmedRetrieval_andBackButton_areInvisible() {
        getTester().startComponentInPage(makePanel());

        getTester().assertInvisible("panel:form:previous");
        getTester().assertInvisible("panel:form:next");

        getTester().assertInvisible("panel:form:pubmedRetrieval");

        getTester().assertInvisible("panel:form:back");
    }

    @Test
    void assertSubmit() {
        getTester().startComponentInPage(makePanel());
        applyTestHackWithNestedMultiPartForms();
        getTester().submitForm("panel:form");
    }

    @Test
    void gettingCallingPage_isNull() {
        SearchablePaperPanel panel = getTester().startComponentInPage(makePanel());
        assertThat(panel.getCallingPage()).isNull();
    }

    @Test
    void isNotAssociatedWithNewsletter() {
        assertThat(makePanel().isAssociatedWithNewsletter()).isFalse();
    }

    @Test
    void isNotAssociatedWithWipNewsletter() {
        assertThat(makePanel().isAssociatedWithWipNewsletter()).isFalse();
    }

    @Test
    void isNotNewsletterInStatusWip() {
        assertThat(makePanel().isaNewsletterInStatusWip()).isFalse();
    }

    @Test
    void modifyNewsletterAssociation_isNoOp() {
        AjaxRequestTarget targetMock = mock(AjaxRequestTarget.class);
        makePanel().modifyNewsletterAssociation(targetMock);
        verifyNoMoreInteractions(targetMock);
    }
}

File: core/core-web/src/test/kotlin/ch/difty/scipamato/core/web/paper/common/SearchablePaperPanelTest.kt
package ch.difty.scipamato.core.web.paper.common

import ch.difty.scipamato.clickLinkSameSite
import ch.difty.scipamato.common.AjaxRequestTargetSpy
import ch.difty.scipamato.core.entity.search.SearchCondition
import de.agilecoders.wicket.core.markup.html.bootstrap.button.BootstrapButton
import de.agilecoders.wicket.extensions.markup.html.bootstrap.form.checkboxx.CheckBoxX
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeNull
import org.apache.wicket.markup.html.form.Form
import org.apache.wicket.model.Model
import org.junit.jupiter.api.Test

@Suppress("SpellCheckingInspection")
internal class SearchablePaperPanelTest : PaperPanelTest<SearchCondition, SearchablePaperPanel>() {

    override fun makePanel(): SearchablePaperPanel = newPanel()

    private fun newPanel(attachments: Boolean? = null, attachmentName: String? = null): SearchablePaperPanel {
        val sc = SearchCondition().apply {
            id = "1"
            number = "100"
            authors = "a"
            firstAuthor = "fa"
            isFirstAuthorOverridden = false
            title = "t"
            location = "l"
            publicationYear = "2017"
            pmId = "pmid"
            doi = "doi"
            createdDisplayValue = "cdv"
            modifiedDisplayValue = "lmdv"
            goals = "g"
            population = "p"
            methods = "m"
            populationPlace = "ppl"
            populationParticipants = "ppa"
            populationDuration = "pd"
            exposurePollutant = "ep"
            exposureAssessment = "ea"
            methodStudyDesign = "msd"
            methodOutcome = "mo"
            methodStatistics = "ms"
            methodConfounders = "mc"
            result = "r"
            intern = "i"
            resultMeasuredOutcome = "rmo"
            resultExposureRange = "rer"
            resultEffectEstimate = "ree"
            conclusion = "cc"
            comment = "c"
            addCode(newC(1, "F"))
            mainCodeOfCodeclass1 = "mcocc1"
            addCode(newC(2, "A"))
            addCode(newC(3, "A"))
            addCode(newC(4, "A"))
            addCode(newC(5, "A"))
            addCode(newC(6, "A"))
            addCode(newC(7, "A"))
            addCode(newC(8, "A"))
            codesExcluded = "1B 2C"
            originalAbstract = "oa"
            hasAttachments = attachments
            attachmentNameMask = attachmentName
        }
        return object : SearchablePaperPanel("panel", Model.of(sc)) {
            private val serialVersionUID: Long = 1L
            override fun onFormSubmit() {
                // no-op
            }

            override fun restartSearchInPaperSearchPage() {
                // no-op
            }

            override fun doOnSubmit() {
                // no-op
            }
        }
    }

    override fun assertSpecificComponents() {
        var b = "panel"
        tester.assertComponent(b, SearchablePaperPanel::class.java)
        assertCommonComponents(b)
        b += ":form"
        assertTextFieldWithLabel("$b:id", "1", "ID")
        assertTextFieldWithLabel("$b:number", "100", "SciPaMaTo-Core-No.")
        assertTextFieldWithLabel("$b:publicationYear", "2017", "Pub. Year")
        assertTextFieldWithLabel("$b:pmId", "pmid", "PMID")
        tester.assertLabel("$b:submit:label", "Search")
        assertTextFieldWithLabel("$b:createdDisplayValue", "cdv", "Created")
        assertTextFieldWithLabel("$b:modifiedDisplayValue", "lmdv", "Last Modified")
        tester.assertComponent("$b:submit", BootstrapButton::class.java)
        verifyCodeAndCodeClassCalls(1)

        tester.clickLinkSameSite("panel:form:tabs:tabs-container:tabs:5:link")
        val bb = "$b:tabs:panel"
        val bbb = "$bb:tab6Form"
        assertTextFieldWithLabel("$bbb:attachmentNameMask", null, "Attachment Name Mask")
        assertComponentWithLabel("$bbb:hasAttachments", CheckBoxX::class.java, null, "W/ or w/o Attachments")

        tester.assertComponent(bbb, Form::class.java)
    }

    @Test
    fun specificFields_areEnabled() {
        tester.startComponentInPage(makePanel())
        tester.isEnabled("panel:form:id")
        tester.isEnabled("panel:form:number")
        tester.isEnabled("panel:form:firstAuthorOverridden")
        tester.isEnabled("panel:form:createdDisplayValue")
        tester.isEnabled("panel:form:modifiedDisplayValue")
    }

    @Test
    fun summary_doesNotExist() {
        tester.startComponentInPage(makePanel())
        tester.assertContainsNot("panel:form:summary")
    }

    @Test
    fun summaryShort_doesNotExist() {
        tester.startComponentInPage(makePanel())
        tester.assertContainsNot("panel:form:summaryShort")
    }

    @Test
    fun navigationButtons_andPubmedRetrieval_andBackButton_areInvisible() {
        tester.startComponentInPage(makePanel())
        tester.assertInvisible("panel:form:previous")
        tester.assertInvisible("panel:form:next")
        tester.assertInvisible("panel:form:pubmedRetrieval")
        tester.assertInvisible("panel:form:back")
    }

    @Test
    fun assertSubmit() {
        tester.startComponentInPage(makePanel())
        applyTestHackWithNestedMultiPartForms()
        tester.submitForm("panel:form")
    }

    @Test
    fun gettingCallingPage_isNull() {
        val panel = tester.startComponentInPage(makePanel())
        panel.callingPage.shouldBeNull()
    }

    @Test
    fun isNotAssociatedWithNewsletter() {
        makePanel().isAssociatedWithNewsletter.shouldBeFalse()
    }

    @Test
    fun isNotAssociatedWithWipNewsletter() {
        makePanel().isAssociatedWithWipNewsletter.shouldBeFalse()
    }

    @Test
    fun isNotNewsletterInStatusWip() {
        makePanel().isaNewsletterInStatusWip().shouldBeFalse()
    }

    @Test
    fun modifyNewsletterAssociation_isNoOp() {
        val targetDummy = AjaxRequestTargetSpy()
        makePanel().modifyNewsletterAssociation(targetDummy)
        targetDummy.components.isEmpty()
        targetDummy.javaScripts.isEmpty()
    }

    @Test
    fun withExcludedCodeFilter() {
        tester.startComponentInPage(newPanel())
        val bbb = prepareCodePanel()
        assertTextFieldWithLabel("$bbb:codesExcluded", "1B 2C", "Excluded Codes")
        verifyCodeAndCodeClassCalls(1)
    }

    @Test
    fun withAttachmentFilter_havingAttachments() {
        tester.startComponentInPage(newPanel(attachments = true))
        val bbb = prepareAttachmentPanel()
        assertTextFieldWithLabel("$bbb:attachmentNameMask", null, "Attachment Name Mask")
        assertComponentWithLabel("$bbb:hasAttachments", CheckBoxX::class.java, true, "W/ or w/o Attachments")
    }

    @Test
    fun withAttachmentFilter_notHavingAttachments() {
        tester.startComponentInPage(newPanel(attachments = false))
        val bbb = prepareAttachmentPanel()
        assertTextFieldWithLabel("$bbb:attachmentNameMask", null, "Attachment Name Mask")
        assertComponentWithLabel("$bbb:hasAttachments", CheckBoxX::class.java, false, "W/ or w/o Attachments")
    }

    @Test
    fun withAttachmentFilter_havingAttachmentFilterWithAttachmentName() {
        tester.startComponentInPage(newPanel(attachmentName = "foo"))
        val bbb = prepareAttachmentPanel()
        assertTextFieldWithLabel("$bbb:attachmentNameMask", "foo", "Attachment Name Mask")
        assertComponentWithLabel("$bbb:hasAttachments", CheckBoxX::class.java, null, "W/ or w/o Attachments")
    }

    private fun prepareCodePanel() = prepareTabPanel(tabIndex = 2)
    private fun prepareAttachmentPanel() = prepareTabPanel(tabIndex = 5)

    private fun prepareTabPanel(tabIndex: Int): String {
        val b = "panel:form:tabs"
        tester.clickLinkSameSite("$b:tabs-container:tabs:$tabIndex:link")
        return "$b:panel:tab${tabIndex + 1}Form"
    }
}
